cpp/src/arrow/scalar.cc

```diff
@@ -153,9 +153,10 @@ struct ScalarHashImpl {
 
   Status ArrayHash(const ArrayData& a) {
     RETURN_NOT_OK(StdHash(a.length) & StdHash(a.GetNullCount()));
-    if (a.buffers[0] != nullptr) {
+    if (a.GetNullCount() != 0 && a.buffers[0] != nullptr) {
```

Comment:
IIRC, it's also possible that `a.buffers[0] == nullptr` if all the elements are valid.  Is it possible that we still get differing hashes in this case?

* All elements valid and equal and validity bitmap present (would hash the validity bitmap)
* All elements valid and equal and validity bitmap missing (would not hash the validity bitmap)

Reply:
Yes, that's the bug that I was trying to fix. Since both cases have the same semantic value, their hashes should be the same.